### DISCUSSION TITLE
[8.0] Migrations files should be no longer loaded in packages

### DISCUSSION THREAD

**Discussion by DarkGhostHunter** [laravel/framework#33495] [General]
I'm having a tantrum here, but I want to put here a bug that has been plaguing me since 5.0.

Basically, when a migration is included in a package, it will be loaded without any problem. But if the user publishes the migration into their `database/migrations`, next time it run `migrate`, PHP will halt saying:

    Cannot declare class CreatePackageTables, because the name is already in use

Having packages load migrations should be discouraged, so users can publish the migration manually and adjust them to their needs.

For this, I think Laravel 8.0 should discourage migration files, and rather, add `publishMigrations` helper to the service provider to allow clean migrations procedures, using the `database/migrations` path as default.

```php
public function boot()
{
    $this->publishMigrations([
        'CreatePackageMigrationsFoo',
        'CreatePackageMigrationsBar',
    ]);
}
```

The above is just syntax sugar to publish the files under the `migrations` tags.

```php
$this->publishes([
    __DIR__ . '/../database/migrations/2020_04_01_000000_create_package_migrations_foo.php' =>
        database_path('migrations/' . now()->format('Y_m_d_His') . '_create_package_migrations_foo.php')
    __DIR__ . '/../database/migrations/2020_04_01_000000_create_package_migrations_bar.php' =>
        database_path('migrations/' . now()->format('Y_m_d_His') . '_create_package_migrations_bar.php')
    ], 'migrations');
```

**Comment by driesvints** [comment]
I don't think we should do this. We've been using package shipped migrations just fine for years (like in Passport). Exporting the migrations and modifying is actually the thing that should be discouraged imo.

**Reply by DarkGhostHunter** [reply]
That makes any migration uneditable.

**Reply by driesvints** [reply]
No it doesn't. You can still publish and edit them if you like. But I would discourage that.
